# Step 1: Import orders.csv dataset

In [1]:
#import libraries
import pandas as pd
import numpy as np
import os

In [2]:
f_path = r'C:\Users\Mei\Instacart Basket Analysis\02 Data'

In [3]:
df_ords = pd.read_csv(os.path.join(f_path,'Original data','orders.csv'), index_col = False)

In [4]:
df_ords.head(5)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [5]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB


In [6]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Step 2: Change the data types of some variables

## change the data type of 'order_id'

In [7]:
df_ords['order_id']=df_ords['order_id'].astype('str')

## change the data type of 'user_id'

In [8]:
df_ords['user_id']=df_ords['user_id'].astype('str')

In [9]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                object 
 1   user_id                 object 
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(3), object(3)
memory usage: 182.7+ MB


# Step3: change the column name without overwriting df_ords

## change the 'oder_dow' to 'order_day_of_week'

In [29]:
df_ords.rename(columns = {'order_dow':'order_day_of_week'}, inplace = False).head(5)

order_id user_id eval_setting  order_number  order_day_of_week  \
0  2539329       1        prior             1                  2   
1  2398795       1        prior             2                  3   
2   473747       1        prior             3                  3   
3  2254736       1        prior             4                  4   
4   431534       1        prior             5                  4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# Step 4: the busiest hour in a day

In [33]:
df_ords['order_hour_of_day'].value_counts(dropna = False).head(10)

10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
Name: order_hour_of_day, dtype: int64

## So the busies hour in a day is 10 o'clock in the moring

# Step 5: defining the meaning of 4 in 'department_id'

In [34]:
df_dep = pd.read_csv(os.path.join(f_path,'Original data','departments.csv'), index_col = False)

In [37]:
df_dep.head(3)

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [38]:
df_dep_t = df_dep.T

In [39]:
df_dep_t.head(3)

0
department_id  department
1                  frozen
2                   other

In [40]:
new_header = df_dep_t.iloc[0]

In [41]:
new_header

0    department
Name: department_id, dtype: object

In [42]:
df_dep_t.head(3)

0
department_id  department
1                  frozen
2                   other

In [43]:
df_dep_new = df_dep_t[1:]

In [48]:
df_dep_new.reset_index().head(3)

department_id index department
0                 1     frozen
1                 2      other
2                 3     bakery

In [46]:
df_dep_new.columns = new_header

In [47]:
df_dep_new.head(3)

department_id department
1                 frozen
2                  other
3                 bakery

In [49]:
data_dict = df_dep_new.to_dict('index')

In [50]:
print(data_dict.get('4'))

{'department': 'produce'}


## So the '4' in df_ords means the 'produce' department

# Step 6: records related to 'breakfast'

In [53]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [55]:
df_prod = pd.read_csv(os.path.join(f_path,'Original data','products.csv'), index_col = False)

In [56]:
df_prod.head(3)

product_id                          product_name  aisle_id  department_id  \
0           1            Chocolate Sandwich Cookies        61             19   
1           2                      All-Seasons Salt       104             13   
2           3  Robust Golden Unsweetened Oolong Tea        94              7   

   prices  
0     5.8  
1     9.3  
2     4.5

In [58]:
df_breakfast = df_prod[df_prod['department_id']==14]

In [59]:
df_breakfast.head(3)

product_id             product_name  aisle_id  department_id  prices
27          28        Wheat Chex Cereal       121             14    10.1
33          34                      NaN       121             14    12.2
67          68  Pancake Mix, Buttermilk       130             14    13.7

## So the file df_breakfast is the file the clients organization wants

# step 7： customers for the dinner parties

In [63]:
var_list =['5','7','12','20']

In [69]:
df_dparty = df_prod.loc[df_prod['department_id'].isin([5,7,12,20])]

In [72]:
df_dparty.head(10)

product_id                                       product_name  aisle_id  \
2            3               Robust Golden Unsweetened Oolong Tea        94   
6            7                     Pure Coconut Water With Orange        98   
9           10     Sparkling Orange Juice & Prickly Pear Beverage       115   
10          11                                  Peach Mango Juice        31   
16          17                                  Rendered Duck Fat        35   
19          20     Pomegranate Cranberry & Aloe Vera Enrich Drink        98   
22          23                             Organic Turkey Burgers        49   
34          35     Italian Herb Porcini Mushrooms Chicken Sausage       106   
38          39           Daily Tangerine Citrus Flavored Beverage        64   
39          40  Beef Hot Links Beef Smoked Sausage With Chile ...       106   

    department_id  prices  
2               7     4.5  
6               7     4.4  
9               7     8.4  
10              7     2.8  
16             12    17.1  
19              7     6.0  
22             12     8.2  
34             12    15.1  
38              7    12.5  
39             12    22.5

## the df_dparty is the information you need

# Step 8: the number of rows of last df

In [73]:
df_dparty.shape

(7650, 5)

## the number of rows is 7650

# Step 9: extract the information for user_id =1

In [74]:
df_user_1 = df_ords[df_ords['user_id']=='1']

In [75]:
df_user_1.head(10)

order_id user_id eval_setting  order_number  order_dow  order_hour_of_day  \
0  2539329       1        prior             1          2                  8   
1  2398795       1        prior             2          3                  7   
2   473747       1        prior             3          3                 12   
3  2254736       1        prior             4          4                  7   
4   431534       1        prior             5          4                 15   
5  3367565       1        prior             6          2                  7   
6   550135       1        prior             7          1                  9   
7  3108588       1        prior             8          1                 14   
8  2295261       1        prior             9          1                 16   
9  2550362       1        prior            10          4                  8   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0

# Step 10: provide details about user_id = 1

## 1) descriptive analysis

In [76]:
df_user_1.describe()

order_number  order_dow  order_hour_of_day  days_since_prior_order
count     11.000000  11.000000          11.000000               10.000000
mean       6.000000   2.636364          10.090909               19.000000
std        3.316625   1.286291           3.477198                9.030811
min        1.000000   1.000000           7.000000                0.000000
25%        3.500000   1.500000           7.500000               14.250000
50%        6.000000   3.000000           8.000000               19.500000
75%        8.500000   4.000000          13.000000               26.250000
max       11.000000   4.000000          16.000000               30.000000

## 2) which day in a week he likes to order the product

In [77]:
df_user_1['order_dow'].value_counts(dropna = False)

4    4
1    3
2    2
3    2
Name: order_dow, dtype: int64

## The fourth day (Thursday) is his favorite day to order products.

## 3) which time in a day he likes to oder things

In [79]:
df_user_1['order_hour_of_day'].value_counts(dropna = False).head(3)

8     3
7     3
12    1
Name: order_hour_of_day, dtype: int64

## He likes order things at 7 and 8 o'clock in the moring

# Step 11: export df_ords as 'orders_wrangled.csv' in 'Prepared Data' folder

In [81]:
df_ords.to_csv(os.path.join(f_path,'Prepared data','orders_wrangled.csv'))

# Step 12: export df_dep_new as 'departments_wrangled.csv' in your 'Prepared data' folder

In [82]:
df_dep_new.to_csv(os.path.join(f_path,'Prepared data','departments_wrangled.csv'))